### PR TITLE
Implement _lstatFile on Windows

### PR DESCRIPTION
**Pull request by jmittert**
Windows doesn't have lstat, but it still can be implemented. This should
help remove some of the Windows specific codepaths.

This is based off the [lstat in llbuild](https://github.com/apple/swift-llbuild/blob/master/lib/Basic/PlatformUtility.cpp#L57)

Also depends on https://github.com/apple/swift/pull/23342

### CODE REVIEW

**Comment by jmittert** [comment]
@compnerd

File: Foundation/FileManager.swift [hunk 1]
```diff
@@ -1818,19 +1817,70 @@ open class FileManager : NSObject {
         }
 
         var statInfo = stat()
+#if os(Windows)
+        let h = path.withCString(encodedAs: UTF16.self) {
+            CreateFileW( /*lpFileName=*/$0,
```
Comment (compnerd) [inline Foundation/FileManager.swift]: Nit: Extra whitespace.

File: Foundation/FileManager.swift [hunk 2]
```diff
@@ -1818,19 +1817,70 @@ open class FileManager : NSObject {
...
+                        /*dwCreationDisposition=*/DWORD(OPEN_EXISTING),
+                        /*dwFlagsAndAttributes=*/DWORD(FILE_FLAG_OPEN_REPARSE_POINT | FILE_FLAG_BACKUP_SEMANTICS),
+                        /*hTemplateFile=*/nil)
+        }
+        if h == INVALID_HANDLE_VALUE {
+            let err = GetLastError()
+            if err == ERROR_FILE_NOT_FOUND {
+                errno = ENOENT
+            }
+            throw _NSErrorWithErrno(errno, reading: true, path: path)
```
Comment (compnerd) [inline Foundation/FileManager.swift]: Just replace this with `_NSErrorWithWindowsError`

File: Foundation/FileManager.swift [hunk 3]
```diff
@@ -1818,19 +1817,70 @@ open class FileManager : NSObject {
...
+        statInfo.st_ino = 0
+        statInfo.st_rdev = info.dwVolumeSerialNumber
+        statInfo.st_mode =
+        // On a symlink to a directory, Windows sets both the REPARSE_POINT and
+        // DIRECTORY attributes. Since Windows doesn't provide S_IFLNK and we
+        // want unix style "symlinks to directories are not directories
+        // themselves, we say symlinks are regular files
+            UInt16((info.dwFileAttributes & DWORD(FILE_ATTRIBUTE_REPARSE_POINT)) != 0
+            ? _S_IFREG
+            : (info.dwFileAttributes & DWORD(FILE_ATTRIBUTE_DIRECTORY)) != 0
```
Comment (compnerd) [inline Foundation/FileManager.swift]: Can you extract the checks into variables please, it makes it easier to read.

Comment (jmittert) [reply Foundation/FileManager.swift]: Done

**Comment by compnerd** [comment]
@swift-ci please test and merge

**Comment by compnerd** [comment]
@swift-ci please test and merge

**Comment by compnerd** [comment]
@swift-ci please test and merge

**Pull request merged**